### PR TITLE
[builder] normalized layer improvements

### LINKED ISSUE #813: feat: the normalized layer should contain gene-length normalized counts from SmartSeq data
The normalized layer is currently not usable for cells coming from SmartSeq assays. SmartSeq assays need to normalize raw counts by gene lengths prior to sum normalization.

When building the normalized layer, Census should be performing gene length normalization for applicable cells.

### LINKED ISSUE #867: [builder] improve float precision reduction on normalized X layer
change precision reduction method to utilize the same method as embeddings, ie., reduce by rounding

### PR DESCRIPTION
**Pull request by bkmartinjr**
Changes:
* For smart-seq based assays, normalized value incorporates feature length.  Fixes #813 
* Improve precision reduction method for normalized layer, in effect storing a PXR24 float. This dramatically reduces error accumulation across the row due to the previous truncation method. Fixes #867 

Note to reviewers: this PR is stacked on top of the schema four PR

### CODE REVIEW

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/experiment_builder.py [hunk 1]
```diff
@@ -361,8 +361,11 @@ def write_X_normalized(self, args: CensusBuildArgs) -> None:
             return
 
         logging.info(f"Write X normalized: {self.name} - starting")
-        # grab the previously calculated sum of the X['raw'] layer
-        raw_sum = self.obs_df.raw_sum.to_numpy()
+        is_smart_seq = np.isin(
+            self.obs_df.assay_ontology_term_id.to_numpy(), ["EFO:0008930", "EFO:0008931", "EFO:0700016"]
```
Comment (ebezzi) [inline tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/experiment_builder.py:365]: These are already defined as a `SMART_SEQ` constant in the code below - use that instead?

**Review by pablo-gar — APPROVED** [review]
Spot checking of the data yield no errors and increased in precision is observed.

**Precision increase**

The min/max row sums of the normalized layer got closer to 1 by aprox 4 orders of magnitude in a small size of census (10K cells)

`min(row_sum_norm)` went from `0.9946784973144531` to `0.9999870657920837`
`max(row_sum_norm)` went from `1.0054092407226562` to `1.0000145435333252`

**No rank order differences between normalized and raw layers**

There is virtually no shuffle of order between normalized and raw layers, min spearman correlation = 0.9999999999999998

**Correct feature-length normalization of smart-seq data**

There are minimal negligible differences between _full_ precision normalization as done 
by Gene Expression (@atarashansky) and the reduced precision as done by Census.

- 98% of values agree at 7 decimal points
- 99.7% of values agree at 6 decimal points

Test data: 

```
"assay == 'Smart-seq2' and dataset_id == 'cee11228-9f0b-4e57-afe2-cfe15ee56312'"
```

```
Decimal points:  10  values differing: n =  2778536 , fraction =  0.754564533509455
Decimal points:  9  values differing: n =  1614282 , fraction =  0.43838911724833146
Decimal points:  8  values differing: n =  515363 , fraction =  0.1399566684336763
Decimal points:  7  values differing: n =  90236 , fraction =  0.024505309719132368
Decimal points:  6  values differing: n =  11112 , fraction =  0.0030176758898776417
Decimal points:  5  values differing: n =  1190 , fraction =  0.0003231672344271413
Decimal points:  4  values differing: n =  142 , fraction =  3.856281284760845e-05
Decimal points:  3  values differing: n =  7 , fraction =  1.9009837319243603e-06
Decimal points:  2  values differing: n =  2 , fraction =  5.431382091212458e-07
Decimal points:  1  values differing: n =  0 , fraction =  0.0
Decimal points:  0  values differing: n =  0 , fraction =  0.0
```

See agreement betwen the two calculations

image

File: tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/experiment_builder.py (hunk 1 above)
Comment (bkmartinjr) [reply tools/cellxgene_census_builder/src/cellxgene_census_builder/build_soma/experiment_builder.py:365]: yes, good catch.  fixing.

**Review by ebezzi — APPROVED** [review]
LGTM!

File: docs/cellxgene_census_schema.md [hunk 2]
```diff
@@ -724,7 +724,9 @@ Per the CELLxGENE dataset schema, [all RNA assays MUST include UMI or read count
 This is an experimental data artifact - it may be removed at any time.
 
 A library-sized normalized layer, containing a normalized variant of the count (raw) matrix.
-For a value `X[i,j]` in the counts (raw) matrix, library-size normalized values are defined
+For Smart-Seq assays, given a value `X[i,j]` in the counts (raw) matrix, library-size normalized values are defined
+as `normalized[i,j] = (X[i,j] / var[j].feature_length) / sum(X[i, ] / var.feature_length[j])`.
+For all other assays, for a value `X[i,j]` in the counts (raw) matrix, library-size normalized values are defined
 as `normalized[i,j] = X[i,j] / sum(X[i, ])`.
```
Comment (atolopko-czi) [inline docs/cellxgene_census_schema.md:730]: Should the schema also mention the precision reduction being performed? Or are we satisfied that it's near equivalence justifies not having to mention it? (I did see Pablo's verification)

**Review by atolopko-czi — APPROVED** [review]
LGTM. (I did _not_ review `validate_soma_bounding_box(), which seems unrelated to this PR).

File: docs/cellxgene_census_schema.md (hunk 2 above)
Comment (bkmartinjr) [reply docs/cellxgene_census_schema.md:730]: this PR does not introduce precision reduction, just improves it.  I leave this up to @pablo-gar to resolve at a later date, at his discretion.

**Pull request merged**